### DISCUSSION TITLE
请问一下插件的点位逻辑是怎样的

### DISCUSSION THREAD

**Discussion by chongchongxiao** [alibaba/ilogtail#1541] [Help] [answered]
以采集文件写到pulsar为例，我理解的逻辑应该是采集文件内容，然后写入到pulsar后，才会更新插件点位，这样才能保证采集的数据内容不会丢失，但是我在flush插件中并没有看到相关的点位处理逻辑，不仅仅是pulsar，其他的插件也没有看到点位相关的处理逻辑，所以想请教一下，这一块大家都是怎么处理的。

**Comment by messixukejia** [comment]
目前是文件读取的时候记录checkpoint，如果主动重启会优先把队列中的清空。后续会计划把checkpoint后置，发送完再记录。

**Comment by chongchongxiao** [comment]
如果还未发送成功，但是点位已经更新了，这个时候因为某种原因进程重启了，那么是不是这一段数据就丢了？这样的话，可靠性咋保证，我理解数据重复还能接受，但是丢了就有点难以接受了。

**Reply by messixukejia** [reply, marked as answer]
![image](https://github.com/alibaba/ilogtail/assets/13250519/035e0b1e-ca31-45de-87bf-c191548315f1)
不会有问题的。
一些原理与测试结果可以看下文章https://mp.weixin.qq.com/s/8mCVk3gvXPOijTlcRjUR_w

**Reply by chongchongxiao** [reply]
主动重启，我也觉得不会有问题，我现在想问的就是异常重启，比如机器宕机或者断电，这个时候记录checkpoint是没有和flush交互的，所以记录checkpoint的时候是无法知道flush插件是否发送成功的，假设这个时候还没有发送成功，但是checkpoint已经被持久化了，这部分还未发送成功的数据不就丢了吗，我实在想不通这块究竟是怎么保证的。你贴的文档我也看了，感觉并没有解答我的问题。

这里一直在说的都是正常结束怎么怎么处理，我想问的是异常情况下，怎么保证数据不丢，我觉得数据重复是可以接受的，但是丢数据有点无法接受。
![image](https://github.com/alibaba/ilogtail/assets/37289259/f46806f0-28e6-42cd-af97-533de3845fcc)

**Reply by messixukejia** [reply]
目前的机制checkpoint会定时记录的、主动重启的时候记录，所以异常重启会最近一次定时的checkpoint开始，一般情况情况下都是会重复的。只有刚记录checkpoint就异常重启这种情况会有些边界情况。后续改成发送后再记录(已经计划上了)，就更精确了。

如果在使用ilogtail，麻烦记录下使用场景吧，多谢https://github.com/alibaba/ilogtail/issues/693

**Comment by chongchongxiao** [comment]
@messixukejia 请问一下，当前这个逻辑是否已经实现
